sklearn/cluster/k_means_.py

```diff
@@ -305,7 +305,7 @@ def k_means(X, n_clusters, init='k-means++', precompute_distances='auto',
         X -= X_mean
 
     if hasattr(init, '__array__'):
-        init = check_array(init, dtype=np.float64, copy=True)
+        init = check_array(init, dtype=X.dtype.type, copy=True)
```

Comment:
(I'm surprised that we need this `.type` here and perhaps `check_array` should allow a dtype object to be passed.)